implementations/rust/c/rust_memory/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+
+use libc::memcmp;
+use std::alloc::Layout;
+use std::mem;
+use std::ptr;
+
+use c_bindings::c_char;
+use c_bindings::{ockam_error_t, ockam_memory_dispatch_table_t, ockam_memory_t};
+
+const ERROR_DOMAIN: &str = "RUST_MEMORY_ERROR_DOMAIN\0";
+
+#[allow(clippy::cast_ptr_alignment)]
+const ERROR_NONE: ockam_error_t = ockam_error_t {
+    code: 0,
+    domain: ERROR_DOMAIN.as_ptr() as *const c_char,
+};
+
+#[repr(C)]
+pub(crate) struct RustAlloc {
+    inner: ockam_memory_t,
+}
+impl AsMut<ockam_memory_t> for RustAlloc {
+    fn as_mut(&mut self) -> &mut ockam_memory_t {
+        &mut self.inner
+    }
+}
+impl RustAlloc {
+    const DISPATCH: ockam_memory_dispatch_table_t = ockam_memory_dispatch_table_t {
+        deinit: Some(self::deinit_impl),
+        alloc_zeroed: Some(self::alloc_zeroed_impl),
+        free: Some(self::free_impl),
+        set: Some(self::memset_impl),
+        copy: Some(self::memcpy_impl),
+        move_: Some(self::memmove_impl),
+        compare: Some(self::memcmp_impl),
+    };
+
+    const GLOBAL: Self = Self {
+        inner: ockam_memory_t {
+            dispatch: &Self::DISPATCH as *const _ as *mut _,
+            context: ptr::null_mut(),
+        },
+    };
+
+    #[inline(always)]
+    pub const fn new() -> &'static Self {
+        &Self::GLOBAL
+    }
+
+    #[inline(always)]
+    pub const fn as_mut_ptr(&self) -> *mut ockam_memory_t {
+        self as *const _ as *mut ockam_memory_t
+    }
+}
+
+unsafe extern "C" fn deinit_impl(_: *mut ockam_memory_t) -> ockam_error_t {
+    ERROR_NONE
+}
+
+unsafe extern "C" fn alloc_zeroed_impl(
+    _: *mut ockam_memory_t,
+    buffer: *mut *mut core::ffi::c_void,
+    size: usize,
+) -> ockam_error_t {
+    let mut error = ERROR_NONE;
+    let layout_result = Layout::from_size_align(size, mem::align_of::<core::ffi::c_void>());
+    if let Ok(layout) = layout_result {
+        let ptr = std::alloc::alloc_zeroed(layout);
+        if !ptr.is_null() {
+            buffer.write(ptr as *mut _);
+            return error;
+        }
+    }
+
+    error.code = -1;
+    error
+}
+
+unsafe extern "C" fn free_impl(
+    _: *mut ockam_memory_t,
+    ptr: *mut core::ffi::c_void,
+    size: usize,
+) -> ockam_error_t {
+    let mut error = ERROR_NONE;
+    let layout_result = Layout::from_size_align(size, mem::align_of::<core::ffi::c_void>());
+    if let Ok(layout) = layout_result {
+        std::alloc::dealloc(ptr as *mut _, layout);
+        error
+    } else {
+        error.code = -1;
+        error
+    }
+}
+
+unsafe extern "C" fn memset_impl(
+    _: *mut ockam_memory_t,
+    ptr: *mut core::ffi::c_void,
+    byte: u8,
+    count: usize,
+) -> ockam_error_t {
+    core::intrinsics::write_bytes(ptr, byte, count);
+    ERROR_NONE
+}
+
+unsafe extern "C" fn memcpy_impl(
+    _: *mut ockam_memory_t,
+    dst: *mut core::ffi::c_void,
+    src: *const core::ffi::c_void,
+    size: usize,
+) -> ockam_error_t {
+    core::intrinsics::copy_nonoverlapping(src, dst, size);
```

Comment:
Is there a reason we're using `core::intrinsics` vs [std::ptr](https://doc.rust-lang.org/beta/std/ptr/index.html)?

`core::instrinsics` only works on nightly and I believe we are targeting stable

Reply:
Changed to std::ptr